cmd/zc_traverser_list.go

```diff
@@ -0,0 +1,91 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/Azure/azure-storage-azcopy/common"
+	"io"
+	"net/url"
+)
+
+// a meta traverser that goes through a list of paths (potentially directory entities) and scans them one by one
+// behaves like a single traverser (basically a "traverser of traverser")
+type listTraverser struct {
+	listReader              io.ReadCloser
+	childTraverserGenerator childTraverserGenerator
+}
+
+type childTraverserGenerator func(childPath string) (resourceTraverser, error)
+
+func (l *listTraverser) traverse(processor objectProcessor, filters []objectFilter) (err error) {
+	// spawn a scanner to read the list of entities one line at a time
+	scanner := bufio.NewScanner(l.listReader)
+	for scanner.Scan() {
+		childPath := scanner.Text()
+
+		// fetch an appropriate traverser, and go through the child path, which could be
+		//   1. a single entity
+		//   2. a directory entity that needs to be scanned
+		childTraverser, err := l.childTraverserGenerator(childPath)
+		if err != nil {
+			glcm.Info(fmt.Sprintf("Skipping %s due to error %s", childPath, err))
+			continue
+		}
+
+		// when scanning a child path under the parent, we need to make sure that the relative paths of
+		// the results are indeed starting right under the parent
+		// ex: parent = /usr/foo
+		// case 1: child1 is a file under the parent
+		//         the relative path returned by the child traverser would be ""
+		//         it should be "child1" instead
+		// case 2: child2 is a directory, and it has items under it such as child2/grandchild1
+		//         the relative path returned by the child traverser would be "grandchild1"
+		//         it should be "child2/grandchild1" instead
+		preProcessor := func(object storedObject) error {
+			if object.relativePath == "" {
```

Comment:
Sorry for the confusion! A better way to think of it (I'll update the comments in zc_enumerator too) is that the rootDir is given by the user, sometimes it turns out the rootDir is actually pointing to a file. So in that case, the relative path is empty since rootDir already points to it, relatively speaking the path is nothing.

rootDir=/var/a/b/c/d/e/f.pdf fullPath=/var/a/b/c/d/e/f.pdf => relativePath=""